core/common/src/main/java/alluxio/resource/SharedResourcePool.java

```diff
@@ -0,0 +1,255 @@
+      } catch (IOException e) {
+        if (ref != null) {
+          // create new resource failed, but can fall back to use
+          // shared resource
+          LOG.warn("Reuse shared resource as creating new resource failed", e);
+          mResources.remove(ref);
+          ref.reference();
+          mResources.add(ref);
+          return makeResource(ref);
+        }
+        // can't fall back, error out
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Wraps the inner {@link CountingReference} in a {@link CloseableResource} for return.
+   * @param ref the inner shared reference of the resource
+   * @return the closeable resource for public use
+   */
+  private CloseableResource<T> makeResource(CountingReference<T> ref) {
+    return new CloseableResource<T>(ref.get()) {
+      @Override
+      public void closeResource() {
+        try (LockResource r = new LockResource(mLock)) {
+          mResources.remove(ref);
+          ref.dereference();
+          mResources.add(ref);
+        }
+      }
+    };
+  }
+
+  /**
+   * @return the number of resources the queue currently holds
+   */
+  public int size() {
+    try (LockResource r = new LockResource(mLock)) {
+      Preconditions.checkState(!mClosed);
+
+      return mResources.size();
+    }
+  }
+
+  /**
+   * @return newly created resource
+   * @throws IOException if error occurs
+   */
+  protected abstract T createNewResource() throws IOException;
+
+  protected void closeResource(T resource) throws IOException {
+    resource.close();
+  }
+
+  /**
+   * Garbage collects unused resources.
+   * @param onlyGcWhenFull if true, then the routine won't do anything if
+   *                       the resource pool hasn't reached full capacity.
+   *                       Otherwise, it gcs all unused resource.
+   */
+  @VisibleForTesting
+  void gcUnusedResources(boolean onlyGcWhenFull) {
+    ArrayList<T> resourceToGc = new ArrayList<>();
+    try (LockResource r = new LockResource(mLock)) {
+      if (onlyGcWhenFull && mResources.size() < mMaxCapacity) {
+        return;
+      }
+      Iterator<CountingReference<T>> iterator = mResources.iterator();
+      while (iterator.hasNext()) {
+        CountingReference<T> nxt = iterator.next();
+        if (nxt.getRefCount() == 0) {
+          resourceToGc.add(nxt.get());
+          iterator.remove();
+        }
+      }
+    }
+
+    for (T resource: resourceToGc) {
+      try {
+        closeResource(resource);
+      } catch (IOException e) {
+        LOG.debug("Resource {} failed to be closed: {}", resource, e);
+      }
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    try (LockResource r = new LockResource(mLock)) {
+      if (mClosed) {
+        return;
+      }
+
+      mClosed = true;
+      int resourceInUse = 0;
+
+      for (CountingReference<T> resource: mResources) {
+        if (resource.getRefCount() != 0) {
+          resourceInUse++;
+        }
+        closeResource(resource.get());
```

Comment:
try and catch exception when one of the resources fails to be closed, so that the for loop can continue to close the rest of the resources in the queue.